cpu/sam0_common/periph/dma.c

```diff
@@ -0,0 +1,334 @@
+    DmacDescriptor *descr = &descriptors[dma];
+    _set_len(descr, len);
+    _set_destination(descr, dst);
+    descr->BTCTRL.reg = (descr->BTCTRL.reg & ~DMAC_BTCTRL_DSTINC) |
+                        (incr << DMAC_BTCTRL_DSTINC_Pos);
+    _set_next_descriptor(descr, NULL);
+}
+
+void _fmt_append(DmacDescriptor *descr, DmacDescriptor *next,
+                 void *src, void *dst, size_t len)
+{
+    /* Configure the full descriptor besides the BTCTRL data */
+    _set_next_descriptor(descr, next);
+    _set_next_descriptor(next, NULL);
+    _set_source(next, src);
+    _set_len(next, len);
+    _set_destination(next, dst);
+}
+
+void dma_append(dma_t dma, DmacDescriptor *next, uint8_t width,
+                void *src, void *dst, size_t len, dma_incr_t incr)
+{
+    DmacDescriptor *descr = &descriptors[dma];
+
+    next->BTCTRL.reg = width << DMAC_BTCTRL_BEATSIZE_Pos |
+                       incr << DMAC_BTCTRL_SRCINC_Pos |
+                       DMAC_BTCTRL_VALID;
+    _fmt_append(descr, next, src, dst, len);
+}
+
+void dma_append_src(dma_t dma, DmacDescriptor *next, void *src, size_t len,
+                    bool incr)
+{
+    DmacDescriptor *descr = &descriptors[dma];
+
+    /* Copy the original descriptor config and modify the increment */
+    next->BTCTRL.reg = (descr->BTCTRL.reg & ~DMAC_BTCTRL_SRCINC) |
+                       (incr << DMAC_BTCTRL_SRCINC_Pos);
+    _fmt_append(descr, next, src, (void *)descr->DSTADDR.reg, len);
+}
+
+void dma_append_dst(dma_t dma, DmacDescriptor *next, void *dst, size_t len,
+                    bool incr)
+{
+    DmacDescriptor *descr = &descriptors[dma];
+
+    /* Copy the original descriptor config and modify the increment */
+    next->BTCTRL.reg = (descr->BTCTRL.reg & ~DMAC_BTCTRL_DSTINC) |
+                       (incr << DMAC_BTCTRL_DSTINC_Pos);
+    _fmt_append(descr, next, (void *)descr->SRCADDR.reg, dst, len);
+}
+
+void dma_start(dma_t dma)
+{
+    DEBUG("[dma]: starting: %u\n", dma);
+
+#ifdef CPU_FAM_SAMD5X
+    DMAC->Channel[dma].CHCTRLA.bit.ENABLE = 1;
+#else
+    unsigned state = irq_disable();
+    DMAC->CHID.bit.ID = dma;
+    DMAC->CHCTRLA.reg = DMAC_CHCTRLA_ENABLE;
+    irq_restore(state);
+#endif
+}
+
+void dma_wait(dma_t dma)
+{
+    DEBUG("[DMA]: mutex lock: %u\n", dma);
+    mutex_lock(&dma_ctx[dma].sync_lock);
+}
+
+void dma_cancel(dma_t dma)
+{
+    DEBUG("[DMA]: Cancelling active transfer: %u\n", dma);
+#ifdef CPU_FAM_SAMD5X
+    DMAC->Channel[dma].CHCTRLA.bit.ENABLE = 0;
+    while (DMAC->Channel[dma].CHCTRLA.bit.ENABLE) {}
+#else
+    unsigned state = irq_disable();
+    DMAC->CHID.bit.ID = dma;
+    /* Write zero to the enable bit */
+    DMAC->CHCTRLA.reg = 0;
+    /* Wait until the active beat is finished */
+    while (DMAC->CHCTRLA.bit.ENABLE) {}
+    irq_restore(state);
+#endif
+}
+
+void isr_dmac(void)
+{
+    uint16_t status = DMAC->INTPEND.reg;
+    dma_t dma = status & DMAC_INTPEND_ID_Msk;
```

Comment:
What happens if multiple interrupts are pending?

Reply:
`DMAC_INTPEND` holds the interrupt status for the highest priority channel with pending interrupts. Interrupts for a specific channel can be cleared by writing the channel ID together with the flags to be cleared to this register. So the interrupt handler only clears interrupts for a single DMA channel and will fire directly again if another channel has pending interrupts.

Reply:
Better add this as a comment then 🙂

Reply:
Done